src/vidpf.rs

```diff
@@ -0,0 +1,644 @@
+            let s_tilde_i_1 = conditional_xor_seeds(&s_diff_1, &s_cw, t_i[1]);
+
+            t_i[0] = t_diff_0 ^ (t_i[0] & t_cw_diff);
+            t_i[1] = t_diff_1 ^ (t_i[1] & t_cw_diff);
+
+            let w_i_0;
+            let w_i_1;
+            (s_i[0], w_i_0) = self.convert(&s_tilde_i_0);
+            (s_i[1], w_i_1) = self.convert(&s_tilde_i_1);
+
+            let mut w_cw =
+                w_i_1 - w_i_0 + Weight::conditional_select(weight, weight, Choice::from(0));
+            w_cw.conditional_negate(t_i[1]);
+
+            let cw_i = CorrectionWord {
+                s_cw,
+                t_cw_l,
+                t_cw_r,
+                w_cw,
+            };
+            cw.push(cw_i);
+
+            let pi_tilde_0 = Self::node_proof(input, level, &s_i[0])?;
+            let pi_tilde_1 = Self::node_proof(input, level, &s_i[1])?;
+            let cs_i = pi_tilde_0.xor(&pi_tilde_1);
+            cs.push(cs_i);
+        }
+
+        Ok(PublicShare { cw, cs })
+    }
+
+    /// [`Vidpf::eval`] evaluates the entire `input` and produces a share of the
+    /// input's weight.
+    ///
+    /// # Errors
+    /// Triggers an error when the public key does not contain as many correction words as
+    /// the length of the input (in bits).
+    pub fn eval(
+        &self,
+        key: &Key,
+        public: &PublicShare<Weight>,
+        input: &VidpfInput,
+    ) -> Result<ValueShare<Weight>, VidpfError> {
+        let mut state = EvalState::init_from_key(key);
+        let mut share = Weight::zero(&self.weight_parameter);
+
+        let n = input.len();
+        for _ in 0..n {
+            (state, share) = self.eval_next(key, public, input, &state)?;
+        }
+
+        Ok(ValueShare {
+            share,
+            proof: state.proof,
+        })
+    }
+
+    /// [`Vidpf::eval_next`] evaluates the `input` at the level indicated by the provided state
+    /// and returns a new state and a share of the input's weight at the next level.
+    pub fn eval_next(
+        &self,
+        key: &Key,
+        public: &PublicShare<Weight>,
+        input: &VidpfInput,
+        state: &EvalState,
+    ) -> Result<(EvalState, Weight), VidpfError> {
+        let level = state.level;
+
+        let CorrectionWord {
+            s_cw,
+            t_cw_l,
+            t_cw_r,
+            w_cw,
+        } = public.cw.get(level).ok_or(VidpfError::IndexLevel)?;
+
+        let seq_tilde = &self.prg(&state.seed);
+
+        let t_i = state.t_control_bit;
+        let sl = conditional_xor_seeds(&seq_tilde.0, s_cw, t_i);
+        let sr = conditional_xor_seeds(&seq_tilde.2, s_cw, t_i);
+        let tl = seq_tilde.1 ^ (t_i & *t_cw_l);
+        let tr = seq_tilde.3 ^ (t_i & *t_cw_r);
+
+        let x_i = Choice::from(u8::from(input[level]));
+        let s_tilde_i = conditional_select_seed(x_i, &[sl, sr]);
+
+        let t_new = Choice::conditional_select(&tl, &tr, x_i);
+        let (seed_new, w_i) = self.convert(&s_tilde_i);
```

Comment:
nit: Align with the function's verbiage
```suggestion
        let next_t = Choice::conditional_select(&tl, &tr, x_i);
        let (next_seed, w_i) = self.convert(&s_tilde_i);
```